lib/closure_tree/hierarchy_maintenance.rb

```diff
@@ -52,59 +52,51 @@ def _ct_after_save
       true # don't cancel anything.
     end
 
-    def _ct_before_destroy
+    def _ct_around_destroy
       _ct.with_advisory_lock(self) do
-        delete_hierarchy_references
+        children_ids = children.pluck(:id).map { |id| _ct.quoted_value(id) }.join(',')
+        delete_references([id])
+
+        yield
+
         if _ct.options[:dependent] == :nullify
-          self.class.find(self.id).children.find_each { |c| c.rebuild! }
+          self.class.where("#{_ct.primary_key_column.name} IN (#{children_ids})").each(&:rebuild!)
```

Comment:
I think that this could be a bug... If `self.class` is part of a single table inheritance hierarchy, do rails filter out the children that do not have the `self.class` type?